src/nativeAssetManager.js

```diff
@@ -330,14 +333,106 @@ export function newNativeAssetManager(win, pubUrl) {
     }
   }
 
+  /** This function returns the element that contains the current iframe. */
+  function getCurrentFrameContainer(win) {
+    let currentWindow = win;
+    let currentParentWindow;
+
+    while (currentWindow !== win.top) {
+        currentParentWindow = currentWindow.parent;
+        if (!currentParentWindow.frames || !currentParentWindow.frames.length) return null;
+        for (let idx = 0; idx < currentParentWindow.frames.length; idx++)
+            if (currentParentWindow.frames[idx] === currentWindow) {
+              if (!currentParentWindow.document) return null; 
+                for (let frameElement of currentParentWindow.document.getElementsByTagName('iframe')) {
+                    if (!frameElement.contentWindow) return null;
+                    if (frameElement.contentWindow === currentWindow) {
+                        return frameElement.parentElement;
+                    }
+                }
+            }
+    }
+}
+
+  function renderAd(html, bid) {
+    // if the current iframe is not a safeframe, try to set the
+    // current iframe width to the width of the container. This
+    // is to handle the case where the native ad is rendered inside
+    // a GAM display ad.
+    if (!envionment.isSafeFrame()) {
+      let iframeContainer = getCurrentFrameContainer(win);
+      if (iframeContainer && iframeContainer.children && iframeContainer.children[0]) {
+        const iframe = iframeContainer.children[0]; 
+        if (iframe.width === '1' && iframe.height === '1') {
+          let width =  iframeContainer.getBoundingClientRect().width;
+          win.document.body.style.width = `${width}px`;
+        }
+      }
+    } else {
+      document.body.style.width = Math.ceil(win.$sf.ext.geom().self.b) + 'px';
+    }
+    win.document.body.innerHTML += html;
+    callback && callback();
+    win.removeEventListener('message', replaceAssets);
+    stopListening();
+    requestHeightResize(bid.adId, (document.body.clientHeight || document.body.offsetHeight), document.body.clientWidth);
+
+    if (typeof window.postRenderAd === 'function') {
+      window.postRenderAd(bid);
+    }
+  }
+
+  function replaceORTBAssetsAndLinks(html, ortb) {
+
+    const getAssetValue = (asset) => {
+      if (asset.img) {
+        return asset.img.url;
+      }
+      if (asset.data) {
+        return asset.data.value;
+      }
+      if (asset.title) {
+        return asset.title.text;
+      }
+      if (asset.video) {
+        return asset.video.vasttag;
+      }
+    }
```

Comment:
this should default to '' (in light of https://github.com/prebid/prebid-universal-creative/pull/171)

Reply:
If I understand well #171 this function should return an empty string when there's no value for that placeholder... correct me if I'm wrong.

Reply:
I added some code after that will check for placeholders of assets that are never returned, so these are removed from the template. The scenario I'm thinking is: the request specifies some optional assets, but these assets are not returned in the response; the publisher wants to remove those from the template - so if the publisher is using a `##hb_native_asset_id_100` and asset with id `100` is not part of the response, it gets deleted.